packages/core/src/services/chains/run.ts

```diff
@@ -305,6 +313,8 @@ async function runStep({
     } else {
       streamConsumer.stepCompleted(response)
 
+      // TODO: Here instead of running the state if the step has
+      // toolCalls we serialize the chain up until this point
```

Comment:
When the AI is cached or when is fresh in both cases the step response can be `stepCompleted` or `chainCompleted` and we're doing the same. 

Now we also want to handle tool calling and chain serialization so I think we need to DRY this in to a new unified method that handle all the cases